server/service/src/preference/types.rs

```diff
@@ -0,0 +1,215 @@
+use repository::{PreferenceRow, RepositoryError, StorageConnection};
+use serde::{de::DeserializeOwned, Serialize};
+
+use thiserror::Error;
+
+use super::{
+    query_preference::{query_global, query_store},
+    upsert_helpers::{upsert_global, upsert_store},
+};
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum PreferenceType {
+    Global,
+    Store,
+    // User,
+    // Machine,
+}
+
+pub enum PreferenceValueType {
+    Boolean,
+    Integer,
+    // String,
+    // Add scalar or custom value types here - mapped to frontend renderers
+}
+
+#[derive(Clone, Error, Debug, PartialEq)]
+pub enum PreferenceError {
+    #[error(transparent)]
+    DatabaseError(RepositoryError),
+    #[error("Failed to deserialize preference {0} from value {1}: {2}")]
+    DeserializeError(String, String, String),
+    #[error("Store ID is required for store preference")]
+    StoreIdNotProvided,
+}
+
+#[derive(Clone, Error, Debug, PartialEq)]
+pub enum UpsertPreferenceError {
+    #[error(transparent)]
+    DatabaseError(RepositoryError),
+    #[error("Failed to serialize preference {0}: {1}")]
+    SerializeError(String, String),
+    #[error("Not running as central server")]
+    NotACentralServer,
+    #[error("Store ID is required for store preference")]
+    StoreIdNotProvided,
+}
+
+pub trait Preference: Sync + Send {
+    type Value: Default + DeserializeOwned + Serialize;
+
+    fn key(&self) -> &'static str;
+
+    fn preference_type(&self) -> PreferenceType;
+
+    fn value_type(&self) -> PreferenceValueType;
+
+    fn query(
+        &self,
+        connection: &StorageConnection,
+        store_id: Option<String>,
+    ) -> Result<Option<PreferenceRow>, PreferenceError> {
+        let pref = match self.preference_type() {
+            PreferenceType::Global => query_global(connection, self.key())?,
+            PreferenceType::Store => {
+                let store_id = store_id.ok_or(PreferenceError::StoreIdNotProvided)?;
+                query_store(connection, self.key(), &store_id)?
+            }
+        };
+
+        Ok(pref)
+    }
+
+    fn default_value(&self) -> Self::Value {
+        Self::Value::default()
+    }
+
+    fn deserialize(&self, data: &str) -> Result<Self::Value, serde_json::Error> {
+        serde_json::from_str::<Self::Value>(data)
+    }
```

Comment:
Exposing methods like deserialize and default as methods on the pref, so we can define custom ones when needed